Improve UIAccessiblityContainer / -accessibilityElements override support

maicki opened the pull request:
- Add provide a way for `ASDisplayNodes` to overwrite `UIAccessibilityContainer` subclass methods
- For `ASDisplayNodes` that acting like a container we provide a default implementation and handling for subnodes
- We add a new experimental flag: `ASExperimentalExposeTextLinksForA11Y` which is by default `NO`
- Provides concrete implementation of the `UIAccessibitliyContainer` methods within ASTextNode2 which based on the `ASExperimentalExposeTextLinksForA11Y` property will return the text as well as all links as accessibility elements

appleguy commented on Source/ASTableView.mm:
```diff
@@ -96,6 +96,16 @@ - (void)didTransitionToState:(UITableViewCellStateMask)state
   [super didTransitionToState:state];
 }
 
+//- (BOOL)isAccessibilityElement
```
Remove this code? Yeah, it does seem likely that we should be relying on UITableView's own implementations anyway.

maicki replied:
Done

maicki replied:
Done

appleguy commented on Source/ASTextNode+Beta.h:
```diff
@@ -35,6 +35,12 @@ NS_ASSUME_NONNULL_BEGIN
+ * A Boolean indicating if the text node will expose links within the attributedText
+ * as UIAccessibitlyElements
+ */
+@property (nonatomic) BOOL accessibilityExposeLinksAsAccessibilityElements;
```
Is there any case where we'd want to disable this? I'd suggest adding it as an ASExperimental flag, and once tested, we can just globally launch it. It should at least be default-on in the end state, and we could add a property later if anyone needs to disable it.

maicki replied:
I think we should gracefully start integrating it with this property of `NO` and run an experiment from outside.

Adlai-Holler replied:
I think we should just bite the bullet and do the experimental flag – in reality it'll be easier & still safe to just announce that "hey in this release links are accessibility elements and we've already vetted the change!" than it will to say "we changed the default/removed the `accessibilityExposeLinksAsAccessibilityElements` flag so be careful." Let's get to the end state fast.

maicki replied:
Just pushed a new commit that will introduce the `ASExperimentalExposeTextLinksForA11Y` flag that is by default `NO` and removed the `accessibilityExposeLinksAsAccessibilityElements` flag.

pinterest-ios commented:
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/5b7ff4f25c8bbafc0f8ebb5f0dbab907/raw/67338f16098f4b8d9f3f3a3f1d17ddffea3e92ce/log.txt)

appleguy commented on Source/Details/_ASDisplayView.mm:
```diff
@@ -78,7 +78,10 @@ @implementation _ASDisplayView
 
   NSArray *_accessibilityElements;
-  CGRect _lastAccessibilityElementsFrame;
+  BOOL _inAccessibilityElementCount;
```
Let's use bitfield packing for all these. Could also combine them with the _inHitTest and _inPointInside options.

maicki replied:
Done

maicki replied:
Done. `_inHitTest` and `_inPointInside`  will happen within #1522

appleguy commented on Source/Private/ASDisplayNode+UIViewBridge.mm:
```diff
@@ -1299,11 +1299,11 @@ - (UIBezierPath *)accessibilityPath
-  _bridge_prologue_read;
-  return _getFromViewOnly(accessibilityElementCount);
-}
+//- (NSInteger)accessibilityElementCount
```
Presumably delete this code, seems like it's now defined in the other category.

maicki replied:
Done

maicki replied:
Done

Adlai-Holler commented on Source/Details/_ASDisplayViewAccessiblity.mm:
```diff
@@ -237,14 +238,19 @@ static void CollectAccessibilityElementsForView(UIView *view, NSMutableArray *el
+  BOOL _inAccessibilityElementCount;
+  BOOL _inIndexOfAccessibilityElement;
+  BOOL _inAccessibilityElementAtIndex;
+  BOOL _inSetAccessibilityElements;
```
Bit field?

maicki replied:
Done

maicki replied:
Done

pinterest-ios commented:
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/243e9ff3459ea100218afcb272d00cc0/raw/b63605a9738a25445b48bbb7bac7f7c58f7da75a/log.txt)

pinterest-ios commented:
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/7408bafe00217ca50b41f320448c5db0/raw/5aed4ecccc3da302ce6805a8bb9c897620ea17ff/log.txt)

pinterest-ios commented:
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/94c438f18a27cc83b90200b9180aad72/raw/25b9ea885d7c04514cf0eea260eb7387960f3382/log.txt)

nguyenhuy commented:
I'll do a thorough review later, but here is some quick notes on the ability to override `accessibilityElements` which is also what #1488 tries to achieve:
- I think it's a great idea to let `-[ASDisplayNode accessibilityElements]` be a point of customization. That is, subclasses can override the method to provide custom elements. Another thing we need to do is make it publicly available. Right now it's declared in `ASDisplayNode+FrameworkPrivate.h`.
- We need to either expose `_accessibilityElements` ivar also (it's in `ASDisplayNodeInternal.h` now), or make it clear in the documentation of `-[ASDisplayNode accessibilityElements]` that clients are responsible for caching the elements themselves, if need to.
- I think there is still a valid use case of the framework to allow customization via a `accessibilityElementsBlock`. For example an app may want to override custom elements of the root node of an `ASViewController` without having to subclass the node. I think we have existing cases in the framework in which we allow both mechanisms for customization. `layoutSpecBlock` and `-layoutSpecThatFits:` is a good example. If we all agree on this, we can coordinate on adding the block after this diff landed. Thoughts? Also, cc @raycsh017.

maicki commented:
@nguyenhuy Addressing your points
- I don't think we should expose `_accessibilityElements` instead of making it clear in the documentation that clients have to do their own caching. This aligns with the behavior and documentation UIKit is stating for overwriting the `UIAccessibilityContainer` methods.
- I don't think we should provide something like `accessibilityElementsBlock` though. One reason I would like to avoid is that it would clutter up the API (`layoutSpecBlock` in theory is great if we would go with composition only, but has it's downsides like subclass support and Texture is still based a lot on subclassing). Imho for now, we should add the most frictionless API based on subclassing, based on the UIKit implementation and documentation.

raycsh017 commented:
Hmm I don't think exposing `accessibilityElements` is too out of line with sticking with the UIKit documentation, since I believe UIKit supports assigning a value to `accessibilityElements` for customization (would be `nil` if not set).

For `accessibilityElementsBlock`, we don't have to add that API, I think the point is really to create a way for people to override `accessibilityElements` without having to subclass. One main use case I see is overriding the `accessibilityElements` of the main node or view on a view controller. While not impossible, it would be pretty annoying to create a subclass to reroute the `accessibilityElements` to focus on other elements (e.g. child view controller's elements).

Ignore this bit if you already considered this, but I've noticed that we invalidate `accessibilityElements` whenever we add a subnode or a subview. We should consider what might be the expected behavior here? I encountered this when I was finding a way to override `accessibilityElements` myself, thought it might be useful to bringing it up here too :)

maicki commented:
@raycsh017 

It's not about exposing `accessibilityElements` it's about exposing `_accessibilityElements` the instance variable. `accessibilityElements` will be exposed.

It's the same with UIKit therefore I would like to not add a new API at least within this PR. To be able to overwrite you have to subclass within UIKit. We could pick that `accessibilityElementsBlock` in a follow up PR. This PR has the goal to add a new foundational API.

I think we should keep that behavior to invalidate the `accessibilityElements` within `addSubview:` / `removeSubview:` for now. Did you run into any problems we may should consider in this case? Also very open for any improvement suggestions :)

pinterest-ios commented:
<!--
  0 Errors
  1 Warning: PR is classed as Work in Progr...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">PR is classed as Work in Progress</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

pinterest-ios commented:
<!--
  0 Errors
  2 Warnings: PR is classed as Work in Progr...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">PR is classed as Work in Progress</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

pinterest-ios commented:
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/21124bf1378367b2868245603e792f08/raw/7cea161c288ba1ca9a7728f2030599179ce4970d/log.txt)

pinterest-ios commented:
<!--
  0 Errors
  2 Warnings: PR is classed as Work in Progr...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">PR is classed as Work in Progress</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

pinterest-ios commented:
<!--
  0 Errors
  2 Warnings: PR is classed as Work in Progr...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">PR is classed as Work in Progress</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

pinterest-ios commented:
<!--
  0 Errors
  2 Warnings: PR is classed as Work in Progr...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">PR is classed as Work in Progress</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

pinterest-ios commented:
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/6ff82b1f7431d33f477c1e7acd48cf5e/raw/3ac191ef2684eb511c91dea9511ce3536393dbf8/log.txt)

pinterest-ios commented:
<!--
  0 Errors
  1 Warning: This is a big PR, please consi...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

Adlai-Holler commented on Source/ASTextNode2.mm:
```diff
@@ -313,6 +330,102 @@ - (UIAccessibilityTraits)defaultAccessibilityTraits
   return UIAccessibilityTraitStaticText;
 }
 
+static void ASUpdateAccessibilityFrame(ASTextNodeAccessiblityElement *accessibilityElement, ASTextLayout *layout, ASDisplayNode * _Nullable containerNode, ASDisplayNode *node) {
```
Like your other functions, add a brief comment summarizing what this function does.

Adlai-Holler commented on Source/ASTextNode2.mm:
```diff
@@ -313,6 +330,102 @@ - (UIAccessibilityTraits)defaultAccessibilityTraits
+    containerNode = containerNode.supernode;
+  }
+  return containerNode;
+}
```
We can keep this but we also already have ASFindClosestViewOfLayer. Since we're on the main thread, traversing the layer hierarchy might make sense.

maicki replied:
We should not touch the view in here as we are also calling it within a layer backed node.

Adlai-Holler replied:
`ASFindClosestViewOfLayer` doesn't touch the view. It finds a view for a given layer.

Adlai-Holler commented on Source/ASTextNode2.mm:
```diff
@@ -313,6 +330,102 @@ - (UIAccessibilityTraits)defaultAccessibilityTraits
+  NSInteger attributedTextLength = attributedText.length;
+  if (attributedTextLength == 0) {
+    _accessibilityElements = accessibilityElements;
+    return _accessibilityElements;
```
Avoid creating an empty mutable array. Set `_accessibilityElements = @[]` here.

Adlai-Holler commented on Source/ASTextNode2.mm:
```diff
@@ -313,6 +330,102 @@ - (UIAccessibilityTraits)defaultAccessibilityTraits
+- (BOOL)isAccessibilityElement {
+  if (ASActivateExperimentalFeature(ASExperimentalTextNode2A11YContainer)) {
+    // Instead of relying on labels accessibility, We implement UIAccessibilityContainer and
+    // handle accessibility with embedded YTNewLabelAccessibilityElements.
```
Hello YouTube! 😛

Adlai-Holler commented on Source/ASTextNode2.mm:
```diff
@@ -412,21 +525,21 @@ - (void)setAttributedText:(NSAttributedString *)attributedText
+  [self setNeedsLayout];
+  
+  // Force display to create renderer with new size and redisplay with new string
+  [self setNeedsDisplay];
```
Why did this get moved? Not a blocker but

Adlai-Holler commented on Source/ASTextNode2.mm:
```diff
@@ -590,9 +703,7 @@ - (id)_linkAttributeValueAtPoint:(CGPoint)point
-  ASTextContainer *containerCopy = [_textContainer copy];
-  containerCopy.size = self.calculatedSize;
-  ASTextLayout *layout = ASTextNodeCompatibleLayoutWithContainerAndText(containerCopy, _attributedText);
+  ASTextLayout *layout = ASTextNodeCompatibleLayoutWithContainerAndText(_textContainer, _attributedText);
```
The comment says the copy is required. Is it not?

Adlai-Holler commented on Tests/ASTextNode2Tests.mm:
```diff
@@ -91,20 +103,205 @@ - (void)testAccessibility
+                [_textNode.accessibilityElements[0] accessibilityLabel], _textNode.accessibilityLabel);
+  XCTAssertTrue([[_textNode.accessibilityElements[0] accessibilityLabel] isEqualToString:_attributedText.string],
+                @"First accessibility element incorrectly returns \n%@\n when it should be \n%@\n",
+                [_textNode.accessibilityElements[0] accessibilityLabel], _textNode.accessibilityLabel);
```
Use XCTAssertEqual and XCTAssertEqualObjects so we get usable errors, throughout this PR.

pinterest-ios commented:
<!--
  0 Errors
  1 Warning: This is a big PR, please consi...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

pinterest-ios commented:
<!--
  0 Errors
  1 Warning: This is a big PR, please consi...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

pinterest-ios commented:
<!--
  0 Errors
  1 Warning: This is a big PR, please consi...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

pinterest-ios commented:
<!--
  0 Errors
  1 Warning: This is a big PR, please consi...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

pinterest-ios commented:
<!--
  0 Errors
  1 Warning: This is a big PR, please consi...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

pinterest-ios commented:
<!--
  0 Errors
  1 Warning: This is a big PR, please consi...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

pinterest-ios commented:
<!--
  0 Errors
  1 Warning: This is a big PR, please consi...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

pinterest-ios commented:
<!--
  0 Errors
  1 Warning: This is a big PR, please consi...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

pinterest-ios commented:
<!--
  0 Errors
  1 Warning: This is a big PR, please consi...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

Adlai-Holler commented on Source/ASTextNode2.mm:
```diff
@@ -315,6 +332,142 @@ - (UIAccessibilityTraits)defaultAccessibilityTraits
+- (NSInteger)accessibilityElementCount
+{
+  if (
+      !ASActivateExperimentalFeature(ASExperimentalTextNode2A11YContainer)) {
```
Blank line

pinterest-ios commented:
<!--
  0 Errors
  1 Warning: This is a big PR, please consi...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

pinterest-ios commented:
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/658427fa29c57591fdbf1b6a76a282af/raw/bed7b2760fa3a8e18f3f22437f85ad455d57e37a/log.txt)

raycsh017 commented:
Ah, we are not making `accessibilityElements` assignable, are we? I think I mistook exposing `accessibilityElements` as making it assignable. 

What I was trying to get at was, I was prototyping making `accessibilityElements` assignable, and found out that the variable gets reset whenever you add a node/view, which wasn't very intuitive and took some time to figure that out.

If that's not the case, never mind. Putting `accessibilityElementsBlock` in a separate PR sounds good 👍

pinterest-ios commented:
<!--
  0 Errors
  1 Warning: This is a big PR, please consi...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

pinterest-ios commented:
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/81366a745b0aa304f3d41a63fe106fe2/raw/2547335e57b63e2e7b3672735657d8a6556cdc94/log.txt)

nguyenhuy commented:
@maicki Can you rebase with master to fix the CI?

maicki commented:
The current state is quit old. Closing for now.
